cunumeric/module.py

```diff
@@ -6227,3 +6227,728 @@ def bincount(
+    #
+    # result.shape = (q_arr.shape, arr.shape -{axis}):
+    #
+    qresult_shape = (*q_arr.shape, *remaining_shape)
+
+    # fails with cunumeric:
+    # (numpy works)
+    #
+    # cunumeric BUG: TypeError: a bytes-like object is required, not 'ndarray';
+    # TODO: open nvbug
+    #
+    # qs_all = ndarray(shape = qresult_shape,
+    #                     buffer = array(result_1D),
+    #                     dtype = to_dtype)
+    #
+    # construct result NdArray, non-flattening approach:
+    #
+    if qs_all is None:
+        qs_all = zeros(qresult_shape, dtype=to_dtype)
+    else:
+        # implicit conversion from to_dtype to qs_all.dtype assumed
+        #
+        if qs_all.shape != qresult_shape:
+            raise ValueError("wrong shape on output array")
+
+    for index, q in np.ndenumerate(q_arr):
+        (gamma, j) = method(q, n)
+        (left_pos, right_pos) = (j, j + 1)
+
+        # (N-1) dimensional ndarray of left, right
+        # neighbor values:
+        #
+        # non-flattening approach:
+        #
+        # extract values at index=left_pos;
+        arr_1D_lvals = arr.take(left_pos, axis)
+        arr_vals_shape = arr_1D_lvals.shape
+
+        if right_pos >= n:
+            # some quantile methods may result in j==(n-1),
+            # hence (j+1) could surpass array boundary;
+            #
+            arr_1D_rvals = zeros(arr_vals_shape, dtype=arr_1D_lvals.dtype)
+        else:
+            # extract values at index=right_pos;
+            arr_1D_rvals = arr.take(right_pos, axis)
+
+        # axis == None, keepdims == False results in arr_1D_*vals.shape = ()
+        # which would result in incorrect in_place_interp();
+        # hence this case needs special treatment;
+        #
+        if (axis is None) and (keepdims is False):
+            if len(index) == 0:
+                qs_all[...] = (
+                    1.0 - gamma
+                ) * arr_1D_lvals + gamma * arr_1D_rvals
+            else:
+                qs_all[index] = (
+                    1.0 - gamma
+                ) * arr_1D_lvals + gamma * arr_1D_rvals
+        else:
+            # vectorized interpolation: the main source of parallelism
+            # (the other one being on `q` inputs; ignored for now)
+            #
+            # non-flattening approach:
+            #
+            if len(index) == 0:
+                # in-place update of qs_all:
+                #
+                in_place_interp(
+                    gamma,
+                    arr_1D_lvals.reshape(qs_all.shape),
+                    arr_1D_rvals.reshape(qs_all.shape),
+                    qs_all,
+                )
+            else:
+                # in-place update of qs_all:
+                #
+                in_place_interp(
+                    gamma,
+                    arr_1D_lvals.reshape(qs_all[index].shape),
+                    arr_1D_rvals.reshape(qs_all[index].shape),
+                    qs_all[index],
+                )
```

Comment:
Combining these two cases like so would simplify the code, and fix this example `print(cn.quantile(cn.arange(3), [0.25, 0.75], axis=0))`, which doesn't produce the right answer currently. Then we can also remove the `in_place_interp` helper function.

```suggestion
        if len(index) == 0:
            qs_all[...] = (
                (1.0 - gamma) * arr_1D_lvals.reshape(qs_all.shape) +
                gamma * arr_1D_rvals.reshape(qs_all.shape)
            )
        else:
            qs_all[index] = (
                (1.0 - gamma) * arr_1D_lvals.reshape(qs_all[index].shape) +
                gamma * arr_1D_rvals.reshape(qs_all[index].shape)
            )
```

Reply:
Done.

Nice find, thanks. Weird, as I tried something _similar_ (just call `in_place_interp()` w/ or w/o `reshape`) and failed. But this works.